src/back/spirv/parser.rs

```diff
@@ -0,0 +1,1545 @@
+    lookup_function: FastHashMap<Word, crate::Handle<crate::Function>>,
+    lookup_function_type: FastHashMap<Word, LookupFunctionType>,
+    lookup_constant: FastHashMap<Word, crate::Handle<crate::Constant>>,
+    lookup_global_variable: FastHashMap<Word, crate::Handle<crate::GlobalVariable>>,
+    lookup_import: FastHashMap<Word, String>,
+    lookup_label: Vec<Word>,
+    current_label: Word,
+}
+
+impl Parser {
+    pub fn new(module: &crate::Module, debug_enabled: bool) -> Self {
+        Parser {
+            module: Module::new(module),
+            id_count: 0,
+            capabilities: vec![],
+            debugs: vec![],
+            annotations: vec![],
+            debug_enabled,
+            void_type: None,
+            lookup_type: FastHashMap::default(),
+            lookup_function: FastHashMap::default(),
+            lookup_function_type: FastHashMap::default(),
+            lookup_constant: FastHashMap::default(),
+            lookup_global_variable: FastHashMap::default(),
+            lookup_import: FastHashMap::default(),
+            lookup_label: vec![],
+            current_label: 0,
+        }
+    }
+
+    fn generate_id(&mut self) -> Word {
+        self.id_count += 1;
+        self.id_count
+    }
+
+    fn bytes_to_words(&self, bytes: &[u8]) -> Vec<Word> {
+        let words: Vec<Word> = bytes
+            .chunks(4)
+            .map(|chars| match chars.len() {
+                4 => {
+                    (chars[3] as u32) << 24
+                        | (chars[2] as u32) << 16
+                        | (chars[1] as u32) << 8
+                        | chars[0] as u32
+                }
+                3 => {
+                    0x0u32 << 24
+                        | (chars[2] as u32) << 16
+                        | (chars[1] as u32) << 8
+                        | (chars[0] as u32) as u32
+                }
+                2 => 0x0u32 << 24 | 0x0u32 << 16 | (chars[1] as u32) << 8 | chars[0] as u32,
+                1 => 0x0u32 << 24 | 0x0u32 << 16 | 0x0u32 << 8 | chars[0] as u32,
+                _ => 0x0u32,
+            })
+            .collect();
+
+        words
+    }
+
+    fn string_to_words(&self, input: &str) -> Vec<Word> {
+        let mut words: Vec<Word> = self.bytes_to_words(input.as_bytes());
+
+        let last_word = words.last().unwrap();
+
+        let last_character = last_word.to_le_bytes()[3];
+
+        if last_character != 0x0 {
+            // nul-termination
+            words.push(0x0u32);
+        }
+
+        words
+    }
+
+    fn instruction_capability(&self, capability: &Capability) -> Instruction {
+        let mut instruction = Instruction::new(Op::Capability);
+        instruction.add_operand(*capability as u32);
+        instruction
+    }
+
+    fn try_add_capabilities(&mut self, capabilities: &[Capability]) {
+        for capability in capabilities.iter() {
+            if !self.capabilities.contains(capability) {
```

Comment:
nit: use `FastHashSet` with a single command for checking and inserting, see [insert](https://doc.rust-lang.org/std/collections/struct.HashSet.html#method.insert) that returns a boolean

Reply:
Okay, yes that sounds good. So this will become:
```rust
fn try_add_capabilities(&mut self, capabilities: &[Capability]) {
    for capability in capabilities.iter() {
        self.capabilities.insert(*capability);
    }
}
```